specs/WebResourceRequested-CustomScheme.md

```diff
@@ -0,0 +1,295 @@
+
+- Are custom scheme URIs considered secure contexts?
+- Which origins should be able to issue requests from these custom schemes to prevent an untrusted origin from reading trusted data from the app?
+
+As a result, we introduce a new API to be able to register custom schemes and allow the end developer to answer these questions as a part of registration.
+
+# Conceptual pages (How To)
+
+WebResourceRequested event can also be raised for custom schemes. For this, the app has to register the custom schemes it would like to be able to issue resource requests for. With each registration, the app will specify whether the URIs with such schemes are considered secure context and it will also need to explicitly specify the origins that are allowed to make requests to these custom scheme URIs to ensure that untrusted origins cannot read trusted data from the app. The registrations are valid throughout the lifetime of the CoreWebView2Environment and browser process and any other CoreWebView2Environments that share the browser process must register exactly same schemes to be able to create a CoreWebView2Environment. The registered custom schemes also participate in [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) and [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP).
+
+For each custom scheme the developer wants to register they can:
+- Set it to be a secure context scheme to prevent insecure content warnings.
+- Provide a list of origins that can make requests to the custom scheme.
+
+# Examples
+
+## CustomSchemeRegistrations
+
+The following code sample registers 2 custom schemes on the CoreWebView2EnvironmentOptions:
+
+- Scheme 1 has https://*.example.com in its allowed origins list.
+- Scheme 2 has a null allowed origins list.
+
+It adds a WebResourceRequested handler for the requests to the custom schemes, which will serve the requests locally. Then it navigates the WebView2 to https://www.example.com and attempts to use XMLHttpRequest (XHR) with both schemes. Only XHR to scheme 1 will succeed because it's the one that has https://www.example.com in its allowed origin list.
+``` c#
+CoreWebView2EnvironmentOptions options = new CoreWebView2EnvironmentOptions();
+string[] allowedOriginsList = new string[1];
+allowedOriginsList[0] = "https://*.example.com";
+
+options.CustomSchemeRegistrations.Add(new CoreWebView2CustomSchemeRegistration("custom-scheme", true /* isSecure */, allowedOriginsList));
+options.CustomSchemeRegistrations.Add(new CoreWebView2CustomSchemeRegistration("custom-scheme-not-in-allowed-origins", true /* isSecure */, null));
```

Comment:
I would prefer `TreatAsSecure` rather than `IsSecure`. Otherwise it _might_ sound like I'm telling the WebView "yes, I want you to make this thing secure!" which isn't what it means at all.

Reply:
Sounds good.

Also give IsSecure default value of false and AllowedOrigins empty.
(And use SchemeName rather than Scheme)